lucene/core/src/java/org/apache/lucene/store/MMapDirectory.java

```
@@ -83,6 +86,19 @@ public class MMapDirectory extends FSDirectory {
    */
   public static final BiPredicate<String, IOContext> NO_FILES = (filename, context) -> false;
 
+  /** Argument for {@link #setGroupingFunction(Function)} that configures no grouping. */
+  public static final Function<String, Optional<String>> NO_GROUPING = filename -> Optional.empty();
+
+  /** Argument for {@link #setGroupingFunction(Function)} that configures grouping by segment. */
+  public static final Function<String, Optional<String>> GROUP_BY_SEGMENT =
+      filename -> {
+        String segmentName = IndexFileNames.parseSegmentName(filename);
+        if (filename.length() == segmentName.length()) {
+          return Optional.empty();
+        }
+        return Optional.of(segmentName);
+      };
```

I think Uwe had floated the idea of doing extra validation that we're actually getting a segment name -- e.g. something like:
```java
String segmentPrefix = IndexFileNames.parseSegmentName(filename);
int segmentPrefixLength = segmentPrefix.length();
if (segmentPrefixLength > 1 && segmentPrefixLength != filename.length() && segmentName.charAt(0) == '_') {
  String segmentName = segmentPrefix.substring(1);
  try {
    // try to parse a proper segment identifier
    Long.parseLong(segmentName, Character.MAX_RADIX);
    return Optional.of(segmentName);
  } catch (NumberFormatException ex) {
    // unable to parse as a segment identifier
  }
}
return Optional.empty();
```

If we can account for all explicit exceptions (where we don't expect to be able to parse a segment name) we could even throw `IllegalArgumentException` here instead of returning `Optional.empty()`? (this would prevent grouping from being used in cases where it should not be, by failing hard -- and could be a good complement to ensuring `READ_ONCE` is used where appropriate).

---

Do we also have a function that parses the Long without reimplementing it here?

---

Maybe add a method to validate and extract a segment name to `IndexFileNames` class. We possibly also have one in test framework that could be moved...

---

I added `IndexFileNames::parseSegmentIdentifier` (from segmentName), which cleans things up nicely.  It still returns an empty optional for non parsable names, but we can provide a test-specific version of the groupingFunction to enforce certain restrictions of our code ( as suggested by @magibney  ). It's probably best to do this in a MockXX test class rather than impose this for all consumers.